LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/DefaultLoRaDataRequestHandler.cs

```diff
@@ -131,31 +132,46 @@ public async Task<LoRaDeviceRequestProcessResult> ProcessRequestAsync(LoRaReques
                             byte[] decryptedPayloadData = null;
                             try
                             {
-                                decryptedPayloadData = loraPayload.GetDecryptedPayload(loRaDevice.AppSKey);
+                                // In case if the mac command is inside the mac payload, we need to decrypt it.
+                                if (loraPayload.GetFPort() == Constants.LORA_FPORT_RESERVED_MAC_MSG)
+                                {
+                                    loraPayload.MacCommands = MacCommand.CreateMacCommandFromBytes(loRaDevice.DevEUI, loraPayload.GetDecryptedPayload(loRaDevice.NwkSKey));
+                                    requiresConfirmation = loraPayload.IsConfirmed() || loraPayload.IsMacAnswerRequired();
+                                }
+                                else
+                                {
+                                    decryptedPayloadData = loraPayload.GetDecryptedPayload(loRaDevice.AppSKey);
+                                }
                             }
                             catch (Exception ex)
                             {
                                 Logger.Log(loRaDevice.DevEUI, $"failed to decrypt message: {ex.Message}", LogLevel.Error);
                             }
 
                             var fportUp = loraPayload.GetFPort();
```

Comment:
I am not sure if I understand correctly. If the Fport is equal to 0, that means the frmpayload is carrying a mac command, therefore I am trying to avoid usage of the decryption. Do you suggest to move the try in the else clause?

Reply:
what I mean here is that you do call "GetFPort()" above. So you could cache that result on the stack and re-use it down here, no?